File: src/typescript/domain/track.ts
import { Message } from "./message";

export default class Track {
  no: number;
  name: string;
  instrumentId: number = 0; // TODO: GUIから選択できるようにする。
  messages: Message[] = [];

  // TODO: conductorTrack(), instrumentalTrack() の用途が限定的なため、Tracks.default() に処理を移動する
  static conductorTrack(): Track{
    const track = new Track(0, `Conductor Track`, 0, []);

    return track;
  }

  static instrumentalTrack(no: number): Track{
    if(no === 0){
      throw new Error("Track No.0 is used as Conductor Track.");
    }

    const track = new Track(no, `Track${no}`, 0, []);
    return track;
  }

  constructor(no: number, name: string, instrumentId: number, messages: Message[]){
    this.no = no;
    this.name = name;
    this.instrumentId = instrumentId;
    this.messages = messages;
  }

  addMessage(message: Message): void {
    this.messages.push(message);
  }

}

export class Tracks {
  tracks: Track[];

  constructor(tracks: Track[]){
    this.tracks = tracks;
  }

  static empty(): Tracks {
    return new Tracks([]);
  }

  static default(): Tracks {
    const tracks = [];

    tracks.push(Track.conductorTrack());
    for(let idx = 1; idx <= 16; idx++){
      const track = Track.instrumentalTrack(idx);
      tracks.push(track);
    }

    return new Tracks(tracks);
  }

  add(no: number): void {
    const track = Track.instrumentalTrack(no);
    this.tracks.push(track);
  }

  get(no: number): Track {
    if(no > this.tracks.length){
      throw new Error("OutOfRangeError.");
    }

    return this.tracks[no];
  }

  addMessage(idx: number, message: Message): void {
    const track = this.get(idx);
    track.addMessage(message);
  }
}

File: src/python/domain/track.py
from mido import MidiTrack, Message
from domain.message import NoteOnMessage
from domain.parser import Parser


class Track:
    def __init__(
        self, no: int, name: str, instrument_id: int, messages: list[NoteOnMessage]
    ) -> None:
        self.no = no
        self.name = name
        self.instrument_id = instrument_id
        self.messages = messages


class MidoTrackHelper:
    @staticmethod
    def mido_instrument_track(track: Track):
        mido_track = MidiTrack()
        mido_track.name = track.name
        for message in Parser.to_mido_messages(track.messages):
            mido_track.append(message)
        return mido_track

    @staticmethod
    def to_sequencer_track(no: int, mido_track: MidiTrack[Message]):
        messages = Parser.to_sequencer_messages(mido_track)
        # TODO: メッセージからinstrument_id を取得できるようにする
        return Track(no, mido_track.name, 0, messages)
